optuna/samplers/nsgaiii.py

```diff
@@ -0,0 +1,643 @@
+from optuna.samplers._base import BaseSampler
+from optuna.samplers._random import RandomSampler
+from optuna.samplers._search_space import IntersectionSearchSpace
+from optuna.samplers.nsgaii._crossover import perform_crossover
+from optuna.samplers.nsgaii._crossovers._base import BaseCrossover
+from optuna.samplers.nsgaii._crossovers._uniform import UniformCrossover
+from optuna.samplers.nsgaii._sampler import _constrained_dominates
+from optuna.study import Study
+from optuna.study._multi_objective import _dominates
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+# Define key names of `Trial.system_attrs`.
+_GENERATION_KEY = "nsga3:generation"
+_POPULATION_CACHE_KEY_PREFIX = "nsga3:population"
+# Define a coefficient for scaling intervals, used in _filter_inf() to replace +-inf
+_COEF = 3
+
+
+class NSGAIIISampler(BaseSampler):
+    """Multi-objective sampler using the NSGA-III algorithm.
+
+    NSGA-III stands for "Nondominated Sorting Genetic Algorithm III",
+    which is a modified version of NSGA-II for many objective optimization problem.
+
+    For further information about NSGA-III, please refer to the following papers:
+
+    - `An Evolutionary Many-Objective Optimization Algorithm Using Reference-Point-Based
+      Nondominated Sorting Approach, Part I: Solving Problems With Box Constraints
+      <https://doi.org/10.1109/TEVC.2013.2281535>`_
+    - `An Evolutionary Many-Objective Optimization Algorithm Using Reference-Point-Based
+      Nondominated Sorting Approach, Part II: Handling Constraints and Extending to an Adaptive
+      Approach
+      <https://doi.org/10.1109/TEVC.2013.2281534>`_
+
+    Args:
+        reference_points:
+            A 2 dimension ``numpy.ndarray`` with objective dimension columns. Represents
+            a list of reference points which is used to determine who to survive.
+            After non-dominated sort, who out of borderline front are going to survived is
+            determined according to how sparse the closest reference point of each individual is.
+            In the default setting the algorithm uses `uniformly` spread points to diversify the
+            result. It is also possible to reflect your `preferences` by giving an arbitrary set of
+            `target` points since the algorithm prioritizes individuals around reference points.
+
+        dividing_parameter:
+            An parameter to determine the density of default reference points. This parameter
+            determines how many divisions are made between reference points on each axis. The
+            smaller this value is, the less reference points you have. The default value is 3.
+            Note that this parameter is not used when reference_points is not None.
+
+    .. note::
+        Other parameters than reference_points and dividing_parameter are the same as
+        :class:`~optuna.samplers.nsgaii.NSGAIISampler`.
+
+    """
```

Comment:
I'm not sure, but the `reference_points` is shown as a public attribute of `NSGAIIISampler`. I don't think it is necessary for users to read/write it.

![image](https://user-images.githubusercontent.com/3255979/234758219-a77e17a0-440d-46e8-b35c-f570e740800d.png)

Reply:
That is because I added getter and setter for `_reference_points` as in the discussion here. https://github.com/optuna/optuna/pull/4436#discussion_r1173189869

Reply:
As we discussed offline, I made `_reference_points` completely private.